### PR TITLE
fix: handle zero inputs in prod backward gradient

### LINKED ISSUE #54551: [Bounty $1,000] ttnn.prod_bw returns non-finite gradients for zero inputs
## Summary

`ttnn.prod_bw` computes the gradient as

```text
prod(input) * grad / input
```

which is `reciprocal(input)` multiplied by a broadcast `prod(input) * grad`. Where the input
contains an exact zero, `reciprocal(0)` is `+inf` and the surviving product is `0`, so the
gradient at that position is non-finite instead of the finite value autograd defines.

```text
input:    [2, 0, 4]
grad:     1
expected: [0, 8, 0]
current:  [0, non-finite, 0]
```

This is a source-level analysis: I do not have a card, so the host replay below renders the
device `inf * 0` as a non-finite value without asserting which one appears in silicon. The defect
is that the gradient is non-finite where the correct value is finite.

## Root cause

Both branches multiply by `reciprocal(input)`:

- `dim=None`:
  [`unary_backward.cpp#L1668-L1676`](https://github.com/tenstorrent/tt-metal/blob/a5cd86212e/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1668-L1676)
  — `multiply(reciprocal(input), fill_first_val_into_tensor(prod * grad))`.
- per-dimension: `reciprocal_input` is built once at
  [`L1711`](https://github.com/tenstorrent/tt-metal/blob/a5cd86212e/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1711)
  and feeds every `dim` branch, e.g. `dim=-1` at
  [`L1722`](https://github.com/tenstorrent/tt-metal/blob/a5cd86212e/ttnn/cpp/ttnn/operations/eltwise/unary_backward/unary_backward.cpp#L1722).
  All four dimensions are affected, not only `dim=-1`.

At a zero input the product of the other values has already been lost in the reduction, so it
cannot be recovered by the division. The formula is valid only when the reduced input has no zeros.

## Reproducer

Host only, no device required (`pip install torch`):

```python
"""Source-level replay of ttnn.prod_bw against torch.prod autograd.

Replays the composite expression from tt-metal main @ a5cd86212e. Runs on host only:
it does not need a Tenstorrent device.
"""

import math
import torch


def source_prod_bw(values, grad):
    """unary_backward.cpp L1668-1676: reciprocal(input) * broadcast(prod(input) * grad)."""
    product = math.prod(values)
    out = []
    for v in values:
        if v == 0.0:
            # reciprocal(0) is +inf on device; inf * 0 is non-finite
            out.append(float("inf") * (product * grad))
        else:
            out.append(product * grad / v)
    return out


def source_prod_bw_dim(rows, grad):
    """unary_backward.cpp L1711-1722, dim=-1: bcast(reciprocal(input), prod * grad, MUL)."""
    return [source_prod_bw(row, grad) for row in rows]


def torch_prod_bw(values, grad):
    t = torch.tensor(values, dtype=torch.float64, requires_grad=True)
    torch.prod(t).backward(torch.tensor(grad, dtype=torch.float64))
    return t.grad.tolist()


def torch_prod_bw_dim(rows, grad):
    t = torch.tensor(rows, dtype=torch.float64, requires_grad=True)
    torch.prod(t, dim=-1, keepdim=True).backward(
        torch.full((len(rows), 1), grad, dtype=torch.float64))
    return t.grad.tolist()


def show(v):
    return "non-finite" if isinstance(v, float) and not math.isfinite(v) else f"{v:g}"


print("dim=None")
for values in ([2.0, 3.0, 4.0], [2.0, 0.0, 4.0], [0.0, 0.0, 4.0], [1.0, 5.0, 0.0]):
    actual, expected = source_prod_bw(values, 1.0), torch_prod_bw(values, 1.0)
    ok = all(a == e for a, e in zip(actual, expected))
    print(f"  input={values}  expected={expected}  "
          f"current=[{', '.join(show(a) for a in actual)}]  {'ok' if ok else 'DIVERGES'}")

print("dim=-1")
for rows in ([[2.0, 3.0, 4.0], [1.0, 5.0, 3.0]], [[2.0, 0.0, 4.0], [1.0, 5.0, 3.0]]):
    actual, expected = source_prod_bw_dim(rows, 1.0), torch_prod_bw_dim(rows, 1.0)
    ok = all(a == e for ar, er in zip(actual, expected) for a, e in zip(ar, er))
    print(f"  input={rows}  expected={expected}  "
          f"current=[{', '.join('[' + ', '.join(show(a) for a in r) + ']' for r in actual)}]  "
          f"{'ok' if ok else 'DIVERGES'}")
```

Output:

```text
dim=None
  input=[2.0, 3.0, 4.0]  expected=[12.0, 8.0, 6.0]  current=[12, 8, 6]  ok
  input=[2.0, 0.0, 4.0]  expected=[0.0, 8.0, 0.0]  current=[0, non-finite, 0]  DIVERGES
  input=[0.0, 0.0, 4.0]  expected=[0.0, 0.0, 0.0]  current=[non-finite, non-finite, 0]  DIVERGES
  input=[1.0, 5.0, 0.0]  expected=[0.0, 0.0, 5.0]  current=[0, 0, non-finite]  DIVERGES
dim=-1
  input=[[2.0, 3.0, 4.0], [1.0, 5.0, 3.0]]  expected=[[12.0, 8.0, 6.0], [15.0, 3.0, 5.0]]  current=[[12, 8, 6], [15, 3, 5]]  ok
  input=[[2.0, 0.0, 4.0], [1.0, 5.0, 3.0]]  expected=[[0.0, 8.0, 0.0], [15.0, 3.0, 5.0]]  current=[[0, non-finite, 0], [15, 3, 5]]  DIVERGES
```

## Expected behaviour

The gradient of a product with respect to one factor is the product of the *other* factors, which
is always finite for finite inputs:

- no zero in the reduced input: the current formula is correct;
- exactly one zero: zero everywhere except at that position, which gets `grad` times the product
  of the non-zero values;
- two or more zeros: zero everywhere.

Suggested tests: exact zeros in both paths, one zero and several zeros, and a non-finite check
before PCC — a PCC comparison alone can pass or silently degrade on a non-finite element.

### PR DESCRIPTION
**Pull request by finsreturns-debug**
## Summary

`ttnn.prod_bw` computes the gradient as `prod(input) * grad * reciprocal(input)`.
When `input` contains an exact zero, `reciprocal(0) = +inf` and the surviving
`prod(input) = 0`, producing a non-finite gradient (`inf * 0`) instead of the
finite value autograd defines.

This PR makes the backward gradient zero-safe by distinguishing the number of
zeros along the reduction dimension:

- 0 zeros:  `prod_result * reciprocal(input)` (unchanged fast path)
- 1 zero:   `prod_nonzero * is_zero` (nonzero only at the zero position)
- >=2 zeros: `0` everywhere

Fixes: #54551

## Verification

⚠️ **Unverified** — I could not build this change.

This is a C++ change (a `ttnn` backward op), which per `AGENTS.md` requires
`.github/scripts/copilot-build.sh`. Docker is not available on my runner
(`docker ABSENT`), and the tt-metal toolchain lives only inside the CI build
image, so I cannot compile or run host-side tests here.

The logic was derived from the mathematical definition
`dy/dx_i = prod_{j != i} x_j` and the reproducer in #54551
(`input=[2,0,4], grad=1 -> expected [0,8,0]`). No accelerator is attached, so
on-device behaviour is also out of scope.

### CODE REVIEW

**Comment by finsreturns-debug** [comment]
Hi @jberkowitzTT, I noticed this PR was closed without a comment or review. Could you share the reason? If the change needs adjustment, I'd be glad to address it. Thanks!

**Comment by jberkowitzTT** [comment]
Hi @finsreturns-debug - the bounty issue #54551 is already assigned. Let me know if this was not in fact an attempt to claim the bounty, let me know!

**Comment by finsreturns-debug** [comment]
Thanks for clarifying @jberkowitzTT. Yes, this PR was an attempt to address the bounty issue #54551. I understand it is already assigned, so I will leave the PR closed. Thanks for your time!
